modules/service/data-service.js

```diff
@@ -91,49 +92,41 @@ class DataService {
         }
     }
 
-    async resolve(assertionId, localQuery = false) {
+    async resolve(id, localQuery = false, metadataOnly = false) {
         try {
-            const rdf = await this.implementation.resolve(assertionId);
-            if (!localQuery && rdf && rdf.find((x) => x.includes(`<did:dkg:${assertionId}> <http://schema.org/hasVisibility> "false"^^<http://www.w3.org/2001/XMLSchema#boolean> .`))) {
+            let {nquads, isAsset} = await this.implementation.resolve(id);
+            if (!localQuery && nquads && nquads.find((x) => x.includes(`<did:dkg:${id}> <http://schema.org/hasVisibility> "private" .`))) {
                 return null;
             }
-            return rdf;
+            if (metadataOnly) {
+                nquads = nquads.filter((x) => x.startsWith(`<did:dkg:${id}>`));
+            }
+            return {nquads, isAsset};
         } catch (e) {
             this.handleUnavailableTripleStoreError(e);
         }
     }
 
-    async createAssertion(assertionId, rdf) {
-        const metadata = rdf.filter((x) => x.startsWith('<did:dkg:'));
-        const data = rdf.filter((x) => !x.startsWith('<did:dkg:'));
-        rdf = rdf.filter((x) => !x.includes('hasKeyword') && !x.includes('hasAsset') && !x.includes('hasRootHash') && !x.includes('hasBlockchain') && !x.includes('hasTransactionHash'));
-        const assertion = await this.implementation.extractMetadata(metadata);
-        assertion.id = assertionId;
-        assertion.data = data;
-        return { assertion, rdf };
+    async createAssertion(nquads) {
+        const metadata = nquads.filter((x) => x.startsWith('<did:dkg:'));
```

Comment:
Why we are calling multiple filters on array. We can do this in only one iteration through array something like this: 
let metadata = []
let data = []
let processedNquads = []
nquads.forEach((nquad)=>{
if (nquad.startsWith('<did:dkg:') then metadata.push(nquad);
else data.push(nquad);
}

Reply:
updated